src/Kiota.Builder/Refiners/TypeScriptRefiner.cs

```diff
@@ -294,4 +296,647 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+    /// </summary>
+    /// <param name="generatedCode"></param>
+    private static void IntroducesInterfacesAndFunctions(CodeElement generatedCode, Func<CodeType, string> functionNameCallback)
+    {
+        CreateSeparateSerializers(generatedCode);
+        CreateInterfaceModels(generatedCode);
+        AddDeserializerUsingToDisriminatorFactory(generatedCode);
+        ReplaceRequestConfigurationsQueryParamsWithInterfaces(generatedCode);
+        AddStaticMethodsUsingsToDeserializerFunctions(generatedCode, functionNameCallback);
+    }
+
+    private static void CreateSeparateSerializers(CodeElement codeElement)
+    {
+        if (codeElement is CodeClass codeClass && codeClass.Kind == CodeClassKind.Model)
+        {
+            CreateSerializationFunctions(codeClass);
+        }
+        CrawlTree(codeElement, x => CreateSeparateSerializers(x));
+    }
+
+    private static void CreateSerializationFunctions(CodeClass modelClass)
+    {
+        var namespaceOfModel = modelClass.GetImmediateParentOfType<CodeNamespace>();
+
+        var serializerMethod = modelClass.Methods.FirstOrDefault(x => x.Kind == CodeMethodKind.Serializer);
+        var deserializerMethod = modelClass.Methods.FirstOrDefault(x => x.Kind == CodeMethodKind.Deserializer);
+        if (serializerMethod == null || deserializerMethod == null)
+        {
+            throw new InvalidOperationException($"The refiner was unable to create local serializer/derserializer method for {modelClass.Name}");
+        }
+        serializerMethod.IsStatic = true;
+        deserializerMethod.IsStatic = true;
+
+        var serializerFunction = new CodeFunction(serializerMethod)
+        {
+            Name = $"{ModelSerializerPrefix}{modelClass.Name.ToFirstCharacterUpperCase()}",
+        };
+
+        var deserializerFunction = new CodeFunction(deserializerMethod)
+        {
+            Name = $"{ModelDeserializerPrefix}{modelClass.Name.ToFirstCharacterUpperCase()}",
+        };
+
+        foreach (var codeUsing in modelClass.Usings.Where(x => x.Declaration != null && x.Declaration.IsExternal))
+        {
+            deserializerFunction.AddUsing(codeUsing);
+            serializerFunction.AddUsing(codeUsing);
+        }
+
+        namespaceOfModel.AddFunction(deserializerFunction);
+        namespaceOfModel.AddFunction(serializerFunction);
+    }
+
+    private static void AddInterfaceParamToSerializer(CodeInterface modelInterface, CodeFunction codeFunction)
+    {
+        var method = codeFunction.OriginalLocalMethod;
+
+        method.AddParameter(new CodeParameter
+        {
+            Name = ReturnFinalInterfaceName(modelInterface.Name), // remove the interface suffix
+            DefaultValue = "{}",
+            Type = new CodeType { Name = ReturnFinalInterfaceName(modelInterface.Name), TypeDefinition = modelInterface }
+        });
+
+        codeFunction.AddUsing(new CodeUsing
+        {
+            Name = modelInterface.Parent?.Name!,
```

Comment:
Optional chaining and asserting a non-null value, this seems very wrong.

It strikes me that this is a case where we'd want to fail to generate, @baywet what's your take here?

Reply:
agreed, we should check whether the parent is not null before trying to add the using, and then the name will not be null

Reply:
@baywet I see a few more instance where we're doing something like 
``` C#
serializer.AddUsing(new CodeUsing
{
    Name = parentSerializer.Parent?.Name!,
    Declaration = new CodeType
    {
        Name = parentSerializer.Name,
        TypeDefinition = parentSerializer
    }
});
```
should I convert everything to first check the parent is null then remove the optional chaining?

``` C#
if(parentSerializer.Parent != null && parentSerializer.Parent.Name != null)
{
	serializer.AddUsing(new CodeUsing
	{
		Name = parentSerializer.Parent?.Name!,
		Declaration = new CodeType
		{
			Name = parentSerializer.Name,
			TypeDefinition = parentSerializer
		}
	});
}
```

Reply:
yes please , however `if(parentSerializer.Parent != null)` should be enough because the name is never null with NRT